# Лабораторная работа №2, Фундаментальные концепции ИИ
### Глобальная оптимизация и метаэврестические алгоритмы
#### Алапанова Эльза, М8О-109СВ-24

В Pygmo запрограммировать две тестовых функции и найти их оптимум 3 разными алгоритмами доступными в библиотеке и получить таблицу сравнения

In [1]:
!pip install pygmo

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.7/14.7 MB 31.4 MB/s eta 0:00:00


In [2]:
import pygmo as pg
import numpy as np
import pandas as pd

Определяем первую тестовую функцию: Sphere

In [3]:
class Sphere:
    def __init__(self, dim=10):
        """Инициализация функции Sphere с заданной размерностью."""
        self.dim = dim

    def fitness(self, x):
        """Целевая функция Sphere: f(x) = sum(x_i^2)."""
        return [sum(xi**2 for xi in x)]

    def get_bounds(self):
        """Границы переменных: [-5.12, 5.12] для каждого измерения."""
        return ([-5.12] * self.dim, [5.12] * self.dim)

Определяем вторую тестовую функцию: Rastrigin

In [4]:
class Rastrigin:
    def __init__(self, dim=10):
        """Инициализация функции Rastrigin с заданной размерностью."""
        self.dim = dim

    def fitness(self, x):
        """Целевая функция Rastrigin: f(x) = 10 * n + sum(x_i^2 - 10 * cos(2 * pi * x_i))."""
        return [10 * len(x) + sum(xi**2 - 10 * np.cos(2 * np.pi * xi) for xi in x)]

    def get_bounds(self):
        """Границы переменных: [-5.12, 5.12] для каждого измерения."""
        return ([-5.12] * self.dim, [5.12] * self.dim)

### Тестовые функции:

Sphere: Простая квадратичная функция.

Rastrigin: Более сложная функция с множеством локальных минимумов.

### Алгоритмы:

Differential Evolution (Дифференциальная эволюция).

Particle Swarm Optimization (Оптимизация роя частиц).

Simulated Annealing (Имитация отжига).



Определяем список алгоритмов оптимизации

In [5]:
algorithms = [
    ("Differential Evolution", pg.algorithm(pg.de(gen=100))),
    ("Particle Swarm Optimization", pg.algorithm(pg.pso(gen=100))),
    ("Simulated Annealing", pg.algorithm(pg.sade(gen=100))),
]

# Список тестовых функций
functions = [Sphere(dim=10), Rastrigin(dim=10)]

# Список для хранения результатов
results = []

In [6]:
# Пробегаем по тестовым функциям и алгоритмам
for func_idx, func in enumerate(functions):
    # Создаем задачу оптимизации для текущей функции
    prob = pg.problem(func)

    for algo_name, algo in algorithms:
        # Создаем начальную популяцию с 20 индивидуумами
        pop = pg.population(prob, size=20)

        # Выполняем эволюцию популяции с использованием текущего алгоритма
        algo.evolve(pop)

        # Сохраняем результаты оптимизации
        results.append({
            "Function": f"Test Function {func_idx + 1}",
            "Algorithm": algo_name,
            "Best Fitness": pop.champion_f[0],  # Лучшая найденная фитнес-значение
            "Best Solution": pop.champion_x     # Координаты лучшего решения
        })

# Формируем таблицу результатов
df_results = pd.DataFrame(results)

# Печатаем таблицу результатов
print(df_results)


          Function                    Algorithm  Best Fitness  \
0  Test Function 1       Differential Evolution     38.798467   
1  Test Function 1  Particle Swarm Optimization     19.264279   
2  Test Function 1          Simulated Annealing     48.314569   
3  Test Function 2       Differential Evolution    104.204745   
4  Test Function 2  Particle Swarm Optimization    148.950940   
5  Test Function 2          Simulated Annealing    122.448606   

                                       Best Solution  
0  [-0.4604350206627217, -1.3794273945411146, 2.3...  
1  [-1.689222310310253, 0.8205636201466575, -1.39...  
2  [-3.505884111016965, 2.21659619097343, 1.61389...  
3  [-1.0699090771694317, 3.3413682010570307, 2.90...  
4  [-2.7190497241213136, 3.471052104654059, -2.00...  
5  [1.5053559557494882, -1.839215392158207, -1.29...  


### Процесс оптимизации

Функция Sphere - простая функция, где глобальный минимум достигается в точке x = (0, 0, ..., 0), а значение функции в этой точке равно 0.

Алгоритмы:\
Differential Evolution (DE):

- Выполняет оптимизацию через комбинацию дифференциальных мутаций и кроссовера.
Полученное значение фитнеса: 38.798467.
- Решение указывает на значительное отклонение от глобального минимума.\

Particle Swarm Optimization (PSO):

- Ищет решение, имитируя поведение роя частиц, движущихся к лучшим найденным позициям.
- Лучшее значение фитнеса: 19.264279, что ближе к минимуму по сравнению с DE.\
PSO лучше справился с задачей за 100 поколений.\

Simulated Annealing (SADE):

- Метод имитирует процесс отжига, постепенно уменьшая вероятность случайных изменений.
- Фитнес: 48.314569, что хуже, чем у DE и PSO.\
Вероятно, SADE недостаточно хорошо справляется с этой функцией в заданных условиях.

Функция Rastrigin: содержит множество локальных минимумов, что делает задачу сложной для алгоритмов глобальной оптимизации.\
Алгоритмы:\

Differential Evolution (DE):
- Найденное значение фитнеса: 104.204745.
Несмотря на сложность функции, DE показал относительно хорошие результаты. Решение оказалось достаточно близким к глобальному минимуму.\

Particle Swarm Optimization (PSO):
- Фитнес: 148.950940.
- Алгоритм PSO продемонстрировал худший результат на функции Rastrigin. Вероятно, он застрял в одном из локальных минимумов.\

Simulated Annealing (SADE):
- Фитнес: 122.448606.
- SADE справился лучше, чем PSO, но всё ещё хуже, чем DE. Это ожидаемо для функции с множеством локальных минимумов.


## Анализ результатов
Функция Sphere:\
- Простая функция без локальных минимумов, но только PSO приблизился к оптимальному решению.
- Simulated Annealing показал худший результат, вероятно, из-за случайных изменений, которые мешали сходимости.\

Функция Rastrigin:
- Для этой сложной функции лучшим алгоритмом оказался Differential Evolution. Его подход на основе мутаций и кроссовера оказался эффективным.
- PSO продемонстрировал слабую способность избегать локальных минимумов.
- Simulated Annealing занял среднее место, что ожидаемо, учитывая стохастический характер алгоритма.



## Вывод

В рамках данной работы была выполнена глобальная оптимизация двух тестовых функций с использованием библиотеки PyGMO. Для каждой функции было применено три различных метаэвристических алгоритма. Результаты экспериментов обобщены в виде таблицы. Для задач с простыми функциями, такими как Sphere, рекомендуется использовать PSO, так как он показал лучшие результаты за то же число итераций.
Для сложных функций с локальными минимумами, таких как Rastrigin, предпочтительно использовать Differential Evolution, так как он показал наилучшие результаты.\
Simulated Annealing менее эффективен в обоих случаях, поэтому его следует применять только для задач, где важна возможность глобального поиска при минимуме итераций.